packages/use-sync-external-store/src/useSyncExternalStore.js

```diff
@@ -7,6 +7,182 @@
  * @flow
  */
 
-export function useSyncExternalStore() {
-  throw new Error('Not yet implemented');
+import * as React from 'react';
+import is from 'shared/objectIs';
+
+// Intentionally not using named imports because Rollup uses dynamic
+// dispatch for CommonJS interop named imports.
+const {
+  useState,
+  useRef,
+  useEffect,
+  useLayoutEffect,
+  useDebugValue,
+
+  // $FlowFixMe - useSyncExternalStore not yet part of React Flow types
+  useSyncExternalStore: builtInAPI,
+} = React;
+
+// Prefer the built-in API, if it exists. If it doesn't exist, then we assume
+// we're in version 16 or 17, so rendering is always synchronous. The shim
+// does not support concurrent rendering, only the built-in API.
+export const useSyncExternalStore =
+  builtInAPI !== undefined ? builtInAPI : useSyncExternalStore_shim;
+
+let didWarnOld18Alpha = false;
+
+// Disclaimer: This shim breaks many of the rules of React, and only works
+// because of a very particular set of implementation details and assumptions
+// — change any one of them and it will break. The most important assumption
+// is that updates are always synchronous, because concurrent rendering is
+// only available in versions of React that also have a built-in
+// useSyncExternalStore API. And we only use this shim when the built-in API
+// does not exist.
+//
+// Do not assume that the clever hacks used by this hook also work in general.
+// The point of this shim is to replace the need for hacks by other libraries.
+function useSyncExternalStore_shim<T>(
+  subscribe: (() => void) => void,
+  getSnapshot: () => T,
+): T {
+  if (__DEV__) {
+    if (!didWarnOld18Alpha) {
+      if (React.startTransition !== undefined) {
+        didWarnOld18Alpha = true;
+        console.error(
+          'You are using an outdated, pre-release alpha of React 18 that ' +
+            'does not support useSyncExternalStore. The ' +
+            'use-sync-external-store shim will not work correctly. Upgrade ' +
+            'to a newer pre-release.',
+        );
+      }
+    }
+  }
+
+  // Because updates are synchronous, we don't queue them. Instead we force a
+  // re-render whenever the subscribed state changes, and read the current state
+  // directly in render.
+  //
+  // We force a re-render by bumping a version number.
+  const [, setVersion] = useState(0);
```

Comment:
If we stored the snapshot in the state, and didn't use it for anything else. That would still enable certain types of bailouts to happen in React itself. Would that lead to the correct behavior? Would it give us any benefits?

It at least avoids the possibility of something updating more than `long` times (highly unlikely). :D

Reply:
Yeah that makes sense. I was thinking we could also enable lazy bailouts by using an inline reducer. If during render we detect that the state hasn’t changed, return the previous state. Otherwise return a new one. Could flip a Boolean back and forth.

Reply:
Or rather, it might bail out incorrectly and not force an update.